src/tactic/declare.lean

```diff
@@ -0,0 +1,194 @@
+(nparams : ℕ)
+(default : list string)
+(mk_cmd : list string → tactic string)
+
+/-- The user attribute for canonical declarations. -/
+@[user_attribute]
+meta def canon_decl_attr : user_attribute (rb_lmap name canon_decl_def) unit := {
+  name := "_canon_decl_attr",
+  descr := "(interal) attribute for canonical declarations",
+  parser := failed,
+  cache_cfg := ⟨λ ns, (do dcls ← ns.mmap (λ n, mk_const n >>= eval_expr (name × canon_decl_def)),
+                          return $ rb_lmap.of_list dcls), []⟩ }
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_canon_decls (n : name) (nparams : ℕ) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls.filter (λ decl, decl.nparams = nparams)
+   end
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_all_canon_decls (n : name) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls
+   end
+
+/-- Execute all commands for the given type of canonical declaration. -/
+@[user_command] meta def declare_cmd (_ : parse $ tk "declare") : parser unit :=
+do do_print ← optional (tk "?"),
+   n ← ident,
+   args ← many ident,
+   let args' := args.map name.to_string,
+   decls ← get_canon_decls n args.length,
+   when (decls.length = 0) $
+     fail format!"no canonical declaration for {n} with {args.length} arguments",
+   decls.mmap' $ λ d, do
+     code ← d.mk_cmd args',
+     when do_print.is_some $ trace ("Try this: " ++ code),
+     emit_code_here code -- is it possible to print the code if this fails?
+
+private meta def parse_declare_string (args : list name) (lst : pexpr) :
+  string → list char → parser pexpr
+| acc [] := pure $ pexpr.of_expr (reflect acc)
+| acc ('%'::'%'::s) := parse_declare_string (acc ++ "%") s
+| acc ('%'::'('::s) :=
+do (e, s) ← with_input ident s.as_string,
+   ')'::s ← return s.to_list | fail "')' expected",
+   f ← parse_declare_string "" s,
+   when (¬ e ∈ args) $ fail format!"no argument {e}",
+   let i := args.index_of e,
+   pure ``(%%(reflect acc) ++ to_string (option.get_or_else (list.nth %%lst %%i) "#error") ++ %%f)
+| acc (c::s) := parse_declare_string (acc.str c) s
+
+/-- Add a new command to a canonical declaration.
+  The new command is added as a declaration to the environment with name `_canon_decl.<number>`.
+  This declaration has attribute `_canon_decl_attr` and as value a `name × canon_decl_def`. -/
+@[user_command] meta def def_declare_cmd (_ : parse $ tk "def_declare") : parser unit :=
+do do_print ← optional (tk "?"),
+   n ← ident,
+   args ← many ident,
+   tk ":=",
+   cmd ← parser.pexpr,
+   cmd ← i_to_expr cmd,
+   cmd ← eval_expr string cmd,
+
+   let v : pexpr := expr.var 0,
```

Comment:
Could you add a few comments to document what's going on here? I didn't appreciate how unhygenic this was (and why it needed to be roundabout like this) until thinking about trying to replace it with `format!`.